ports/coreMQTT/azure_iot_core_mqtt.c

```diff
@@ -0,0 +1,233 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved. */
+/* SPDX-License-Identifier: MIT */
+
+/**
+ * @file azure_iot_core_mqtt.c
+ * @brief Implements the port for Azure IoT MQTT based on coreMQTT.
+ * 
+ */
+ 
+#include <assert.h>
+
+#include "azure_iot_mqtt.h"
+
+/**
+ * Maps CoreMQTT errors to AzureIoTMQTT errors.
+ **/
+static AzureIoTMQTTResult_t prvGetAzureIoTMQTTResult( MQTTStatus_t xResult )
+{
+    AzureIoTMQTTResult_t xReturn;
+
+    switch ( xResult )
+    {
+        case MQTTSuccess :
+        {
+            xReturn = eAzureIoTMQTTSuccess;
+        }
+        break;
+
+        case MQTTBadParameter :
+        {
+            xReturn = eAzureIoTMQTTBadParameter;
+        }
+        break;
+
+        case MQTTNoMemory :
+        {
+            xReturn = eAzureIoTMQTTNoMemory;
+        }
+        break;
+
+        case MQTTSendFailed:
+        {
+            xReturn = eAzureIoTMQTTSendFailed;
+        }
+        break;
+
+        case MQTTRecvFailed :
+        {
+            xReturn = eAzureIoTMQTTRecvFailed;
+        }
+        break;
+
+        case MQTTBadResponse :
+        {
+            xReturn = eAzureIoTMQTTBadResponse;
+        }
+        break;
+
+        case MQTTServerRefused :
+        {
+            xReturn = eAzureIoTMQTTServerRefused;
+        }
+        break;
+
+        case MQTTNoDataAvailable :
+        {
+            xReturn = eAzureIoTMQTTNoDataAvailable;
+        }
+        break;
+
+        case MQTTIllegalState :
+        {
+            xReturn = eAzureIoTMQTTIllegalState;
+        }
+        break;
+
+        case MQTTStateCollision :
+        {
+            xReturn = eAzureIoTMQTTStateCollision;
+        }
+        break;
+
+        case MQTTKeepAliveTimeout :
+        {
+            xReturn = eAzureIoTMQTTKeepAliveTimeout;
+        }
+        break;
+
+        default :
+        {
+            xReturn = eAzureIoTMQTTFailed;
+        }
+        break;
+    }
+
+    return xReturn;
+}
+
+AzureIoTMQTTResult_t AzureIoTMQTT_Init( AzureIoTMQTTHandle_t xContext,
+                                        const AzureIoTTransportInterface_t * pxTransportInterface,
+                                        AzureIoTMQTTGetCurrentTimeFunc_t xGetTimeFunction,
+                                        AzureIoTMQTTEventCallback_t xUserCallback,
+                                        uint8_t * pucNetworkBuffer,
+                                        uint32_t ulNetworkBufferLength )
+{
+    MQTTFixedBuffer_t xBuffer = { pucNetworkBuffer, ulNetworkBufferLength };
+    MQTTStatus_t xResult;
+
+    /* Check memory equivalence, but ordering is not guarantee */
+    assert( sizeof( AzureIoTMQTTConnectInfo_t ) == sizeof( MQTTConnectInfo_t ) );
```

Comment:
Is there a compile time, non-hacky, C99 friendly way to check this assert ala C++ `CASSERT`.  All my requirements around this make me think answer is no :), but in perfect world if this was wrong it'd be a compile error.

Reply:
```
#if sizeof( AzureIoTMQTTConnectInfo_t ) != sizeof( MQTTConnectInfo_t )
#error Only Debug builds are supported
#endif
```

Reply:
Discuss offline.

Possible solution can be to define a macro:

#define azureiotSTATIC_ASSERT(COND,MSG) typedef char static_assertion_##MSG[(COND)?1:-1]


but the error generated by this marco, is not useful.

Reply:
#ignore